node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1997,73 +1973,6 @@ fn verify_trusted_block_header<REv>(ctx: &ChainSyncContext<'_, REv>) -> Result<(
     Ok(())
 }
 
-/// Returns `Ok(Some(FastSyncOutcome::ShouldCommitUpgrade))` if we should commit an upgrade before
-/// syncing further, or `Ok(None)` if not.
-///
-/// If this returns `Ok(Some...)`, we sync the trie store in preparation for running commit_upgrade.
-async fn prepare_for_upgrade_if_needed<REv>(
```

Comment:
Since the PR is about creating an immediate switch block in the _initializer_ reactor rather than _joiner_, I'm surprised that "preparing for upgrade" is being removed as part of this PR.

Reply:
This is because the upgrades will no longer be relevant to the joiner. After the support for trusted hashes from before the last upgrade has been removed, the last case when the joiner could be advancing through an upgrade boundary was when the node was starting right after an upgrade and the highest block in storage was the one right before the upgrade. Once we create the immediate switch block in the initializer, there will be no such cases left and the joiner will always be able to assume no upgrade boundaries when advancing forward from the trusted hash / highest block in storage.

Reply:
Since joining can still take a while, is it possible to stage an upgrade while it's ongoing?

Reply:
Hmm... maybe :thinking:

Reply:
I think we do already handle that - if we fetch a block from a future version we exit the fast sync task with `Error::RetrievedBlockHeaderFromFutureVersion` and that's handled in the component bu setting `self.joining_outcome = Some(JoiningOutcome::ShouldExitForUpgrade);`